### PR TITLE
Smooth throttle in navigation modes on FW

### PR DESCRIPTION
**Pull request by Airwide**
This pull request resolves issue #6073.

**Usage**
Set new cli variable `nav_fw_pitch2thr_threshold` to a value between 0 and 100 decidegrees. That equals 0 to 10 degrees. 0 is default and keeps current behaviour.

The variable is also accessible in the OSD menu under features/navigation/fixed wing/.

**Testing**
You can see in the movies that the red throttle curve is more stable in the video with `nav_fw_pitch2thr_threshold = 30` than in the video with `nav_fw_pitch2thr_threshold = 0`.

`nav_fw_pitch2thr_threshold = 0`:
nav_fw_pitch2thr_threshold_0

`nav_fw_pitch2thr_threshold = 30`:
nav_fw_pitch2thr_threshold_30

Btw, the higher throttle and speed with `nav_fw_pitch2thr_threshold = 0` indicates to me that my board orientation is off by a degree or two.

@avs2 - Can you verify this is what you requested?

### CODE REVIEW

**Comment by avs2** [comment]
Perfect, exactly what I had in mind. Thanks!

**Comment by avs2** [comment]
I think you are not supposed to change `Settings.md` since it is auto generated.

**Comment by Airwide** [comment]
> I think you are not supposed to change `Settings.md` since it is auto generated.

I'm not so sure. I think the auto generated note is actually a note for the table on the web page [Settings.md](https://github.com/iNavFlight/inav/blob/master/docs/Settings.md):

image

I could be wrong though.

**Comment by Airwide** [comment]
@digitalentity What do you think of this pull request? Anything else I need to add before it can be merged?

**Comment by avs2** [comment]
I have just build and flashed this for testing on my Dart 250. I will try to fly it later today and let you know how it went. Looks promising.

**Comment by avs2** [comment]
I tested this (together with #6101) and I think it works fine. Throttle level seems to be more stable. It was *way* too windy for this little wing so it was hard to see the exact effect. I have to say that I did get some substantial horizon drift and the OSD altitude fluctuated by +-20 meters when turning the nose or tail into the wind direction. I think it's unlikely that this is related to this PR but I want to mention it anyway for completeness.

**Comment by avs2** [comment]
This PR calms down the throttle behavior around zero degrees pitch, but outside of ±`nav_fw_pitch2thr_threshold`/10 degrees pitch throttle can fluctuate as before. Also, throttle will stay closer to cruise throttle level if `nav_fw_pitch2thr` is left unchanged. Would it be difficult to apply a filter to the pitch angle to smooth out the throttle behavior for all pitch angles? I think a simple moving average would do, but this firmware already includes some way more advanced filters so there may be better options.

**Comment by Airwide** [comment]
@avs2 - Interesting idea. Let's say we have 50% cruise throttle. I'm thinking that at 5 degrees filtered (or projected) ascent you would like the deadband to be centered around 55% and at 45% for a 5 degrees descent. Let me see what I can come up with.

**Comment by avs2** [comment]
@Airwide Hmm, so you're saying to keep the throttle constant at `nav_fw_cruise_throttle` - "`target_angle`" * `nav_fw_pitch2thr` as long as "`actual_angle`" is within `nav_fw_pitch2thr_threshold`/10 degrees of "`target_angle`"? I wonder if this will have unexpected effects when "`target_angle`" is changing.

**Review by digitalentity — DISMISSED** [review]

**Comment by Airwide** [comment]
@avs2 You're right, using the target angle would cause weird throttling in the target transitions. A moving average (or filtered) angle throttle compensation would work much better.

**Comment by Airwide** [comment]
Ok, I have added `const int16_t baseThrottleCorrection`which is calculated from the 128 last pitch values (`const int16_t movingAverageCycles = 128`)

I thought it was unnecessary to create a cli variable for `const int16_t movingAverageCycles = 128`. The question is what the constant should be set to. How often is `fixedWingPitchToThrottleCorrection(int16_t pitch)` called? Every loop?

@avs2, @digitalentity WDYT?

File: src/main/navigation/navigation_fixedwing.c [hunk 1]
```diff
@@ -463,17 +463,20 @@ int16_t applyFixedWingMinSpeedController(timeUs_t currentTimeUs)
 
 int16_t fixedWingPitchToThrottleCorrection(int16_t pitch)
 {
-    if (pitch > navConfig()->fw.pitch_to_throttle_thresh) {
-        // Positive pitch above threshold
-        return DECIDEGREES_TO_DEGREES(pitch - navConfig()->fw.pitch_to_throttle_thresh) * navConfig()->fw.pitch_to_throttle;
+    // Calculate base throttle correction from pitch moving average
+    const int16_t movingAverageCycles = 128; //Number of main loop cycles for average calculation
```
Comment (avsaase) [inline src/main/navigation/navigation_fixedwing.c]: I think the moving average window should be defined in terms of a unit of time instead of loop cycles, so that the amount of smoothing is consistent across loop frequency settings. Unless there is some amount of smoothing that will work for everyone, and @digitalentity is okay with changing the default behavior, I think there should be a CLI command to set the moving average window time.

File: src/main/navigation/navigation_fixedwing.c [hunk 2]
```diff
@@ -463,17 +463,20 @@ int16_t applyFixedWingMinSpeedController(timeUs_t currentTimeUs)
...
+    static int16_t averagePitch = (averagePitch * movingAverageCycles + pitch - averagePitch) / movingAverageCycles;
+    const int16_t baseThrottleCorrection = DECIDEGREES_TO_DEGREES(averagePitch) * navConfig()->fw.pitch_to_throttle;
```
Comment (avsaase) [inline src/main/navigation/navigation_fixedwing.c]: This line already accomplishes what I had in mind for using a smoothed pitch angle for the throttle correction input.

File: src/main/navigation/navigation_fixedwing.c [hunk 3]
```diff
@@ -463,17 +463,20 @@ int16_t applyFixedWingMinSpeedController(timeUs_t currentTimeUs)
...
+
+    // Calculate final throttle correction
```
Comment (avs2) [inline src/main/navigation/navigation_fixedwing.c:473]: Not entirely sure what this block does. Is it to apply an extra throttle correction based on the difference between `pitch` and `averagePitch`? Say the airplane is pitched up 7 degrees (`pitch=-70`), with the average pitch being up 3 degrees (`averagePitch=-30`), then this subtracts 70 us from the throttle signal calculated above? Or take the case when `nav_fw_pitch2thr_threshold=0`, then this adds (`pitch` - `averagePitch`)*`fw.pitch_to_throttle` to the throttle value.

Can you explain the goal of this a bit?

**Comment by avs2** [comment]
Nice! I have added some comments to the code.

File: src/main/navigation/navigation_fixedwing.c [hunk 4]
```diff
@@ -463,7 +463,21 @@ int16_t applyFixedWingMinSpeedController(timeUs_t currentTimeUs)
...
+    if (pitch > (averagePitch + navConfig()->fw.pitch_to_throttle_thresh)) {
+        return baseThrottleCorrection + DECIDEGREES_TO_DEGREES(pitch - averagePitch - navConfig()->fw.pitch_to_throttle_thresh) * navConfig()->fw.pitch_to_throttle;
+    }
+    else if (pitch < (averagePitch - navConfig()->fw.pitch_to_throttle_thresh)) {
+        return baseThrottleCorrection + DECIDEGREES_TO_DEGREES(pitch - averagePitch + navConfig()->fw.pitch_to_throttle_thresh) * navConfig()->fw.pitch_to_throttle;
+    }
+    else {
+        return baseThrottleCorrection;
+    }
```
Comment (avsaase) [inline src/main/navigation/navigation_fixedwing.c]: Not entirely sure what this block does. Is it to apply an extra throttle correction based on the difference between `pitch` and `averagePitch`? Say the airplane is pitched up 7 degrees (`pitch=-70`), with the average pitch being up 3 degrees (`averagePitch=-30`), then this subtracts 70 us from the throttle signal calculated above? Or take the case when `nav_fw_pitch2thr_threshold=0`, then this adds (`pitch` - `averagePitch`)*`fw.pitch_to_throttle` to the throttle value.

Can you explain the goal of this a bit?

File: src/main/navigation/navigation_fixedwing.c (hunk 1 above)
Comment (Airwide) [reply src/main/navigation/navigation_fixedwing.c]: I see your point about using time for the moving average window. The reason I kept it as a "binary" value is that it is very efficient to just left shift bits for division. Maybe that doesn't matter. @digitalentity What's your thought on this?

Regarding CLI command, I agree. Let me create a CLI variable to keep the current behaviour as default.

File: src/main/navigation/navigation_fixedwing.c (hunk 4 above)
Comment (Airwide) [reply src/main/navigation/navigation_fixedwing.c]: This section creates a deadband (+/-`nav_fw_pitch2thr_threshold`) around `averagePitch`, where the throttle correction will be kept stable at `baseThrottleCorrection`. By setting `nav_fw_pitch2thr_threshold=0`, you basically disable that deadband.

**Comment by Airwide** [comment]
> Nice! I have added some comments to the code. Have you tested this already? For testing purposes it would be nice to have a CLI command so that you can quickly turn the smoothing on and off in the field.

I have not had the opportunity to test. My AR Wing crashed a couple of weeks ago, caused by the FrSky failsafe bug disarming the wing mid flight. Hopefully I will be up in the air again in the upcoming weeks.

File: src/main/navigation/navigation_fixedwing.c (hunk 4 above)
Comment (avsaase) [reply src/main/navigation/navigation_fixedwing.c]: Just to make sure I understand, the idea is that `averagePitch` is equal or close to the target pitch angle (zero for level flight, etc), and at that level of pitch throttle is calculated similar to how it was before this PR. Deviations from that average larger than `fw.pitch_to_throttle_thresh` cause an additional throttle correction. Is this second correction to allow for a more direct response to sudden required pitch changes instead of waiting until `averagePitch` picks up the change?

Just thinking through what happens when you are flying level (`pitch`=`averagePitch`=0) and suddenly get into sinking air. You lose some altitude and the airplane pitches up to correct. If only a small amount of pitch up is required throttle stays at cruise level, because `averagePitch` has barely changed. If more pitch up is needed throttle is immediately increased to avoid losing air speed. When you keep flying through sinking air the airplane will continue to pitch up to maintain altitude so that `averagePitch` accumulates all the pitch angle so `averagePitch`=`pitch` and throttle is managed as in the level flight case. Correct?

It sounds very clever, but I think this needs to be tested in various conditions to avoid overlooking some edge case where this behavior can cause problems. As long is as `movingAverageCycles` is not too large compared to the update frequency of `fixedWingPitchToThrottleCorrection` (no idea what that is btw) I don't foresee and issues. I would like to experiment with some longer moving average windows such as 1 second to see how that behaves.

Awesome work!

**Comment by Airwide** [comment]
This pull request resolves issue #6073.

This commit has two new variables `nav_fw_pitch2thr_smoothing` and `nav_fw_pitch2thr_threshold`, which together with `nav_fw_pitch2thr`should make it possible to tune in a stable throttle for maximum efficiency during wp missions and still keep the airspeed stable at different targeted pitch angles.

## Usage
Latest commit has 2 new variables:

### 1. nav_fw_pitch2thr_smoothing
This variable holds the number of fixedWingPitchToThrottleCorrection cycles used to calculate a moving average for the pitch angle which is used for the pitch to throttle correction. Set `nav_fw_pitch2thr_smoothing` to a value between 1 and 4096. To keep the smoothing calculation efficient, choose a value for `nav_fw_pitch2thr_smoothing` that equals a power of 2 (1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096). 1 is default and keeps current behaviour with no smoothing. 

`nav_fw_pitch2thr_smoothing` is also accessible in the OSD menu under features/navigation/fixed wing/.

### 2. nav_fw_pitch2thr_threshold
This variable creates a deadband of +/- `nav_fw_pitch2thr_threshold` around the average pitch.
Set `nav_fw_pitch2thr_threshold` to a value between 0 and 100 centidegrees. That equals 0 to 10 degrees.  0 is default and keeps current behaviour.

`nav_fw_pitch2thr_threshold` is also accessible in the OSD menu under features/navigation/fixed wing/.

## Testing
I'm currently grounded, and haven't been able to test this latest commit. I should be up in the air in the next week or so. Meanwhile, help with testing and feedback would be much appreciated. Let me know if you have questions.

File: src/main/navigation/navigation_fixedwing.c (hunk 1 above)
Comment (Airwide) [reply src/main/navigation/navigation_fixedwing.c]: I have now added `nav_fw_pitch2thr_smoothing` to the cli settings, but kept it as a cycle count for the time being.

File: src/main/navigation/navigation_fixedwing.c (hunk 4 above)
Comment (Airwide) [reply src/main/navigation/navigation_fixedwing.c]: > Just to make sure I understand, the idea is that averagePitch is equal or close to the target pitch angle (zero for level flight, etc), and at that level of pitch throttle is calculated similar to how it was before this PR. Deviations from that average larger than fw.pitch_to_throttle_thresh cause an additional throttle correction. Is this second correction to allow for a more direct response to sudden required pitch changes instead of waiting until averagePitch picks up the change?

Correct!

>Just thinking through what happens when you are flying level (pitch=averagePitch=0) and suddenly get into sinking air. You lose some altitude and the airplane pitches up to correct. If only a small amount of pitch up is required throttle stays at cruise level, because averagePitch has barely changed. If more pitch up is needed throttle is immediately increased to avoid losing air speed. When you keep flying through sinking air the airplane will continue to pitch up to maintain altitude so that averagePitch accumulates all the pitch angle so averagePitch=pitch and throttle is managed as in the level flight case. Correct?

Correct, and this also applies to the case where you have an altitude change between two waypoints.

>I would like to experiment with some longer moving average windows such as 1 second to see how that behaves.

I'm even thinking 2 to 5 seconds and then tune `nav_fw_pitch2thr_threshold` to get a balance between transition response and smooth throttle in between.

**Comment by avs2** [comment]
@b14ckyy, you requested this change as well. Would you be willing to test it?

**Comment by b14ckyy** [comment]
Yeah I just saw the update and to test it was my first thought. As it is not yet merged to master, can you build me a OMNIBUSF4V6 Target please? Then I can test it on the weekend on my test platform. I just learned how to build my own version from the master branch but I have no idea how I can build one with a not merged change.

**Comment by Airwide** [comment]
> @avs2 Yeah I just saw the update and to test it was my first thought. As it is not yet merged to master, can you build me a OMNIBUSF4V6 Target please? Then I can test it on the weekend on my test platform. I just learned how to build my own version from the master branch but I have no idea how I can build one with a not merged change.

Great! Here's you go.
[inav_2.6.0_OMNIBUSF4V6.hex.zip](https://github.com/iNavFlight/inav/files/5306266/inav_2.6.0_OMNIBUSF4V6.hex.zip)

**Comment by b14ckyy** [comment]
> > @avs2 Yeah I just saw the update and to test it was my first thought. As it is not yet merged to master, can you build me a OMNIBUSF4V6 Target please? Then I can test it on the weekend on my test platform. I just learned how to build my own version from the master branch but I have no idea how I can build one with a not merged change.
> 
> Great! Here's you go.
> [inav_2.6.0_OMNIBUSF4V6.hex.zip](https://github.com/iNavFlight/inav/files/5306266/inav_2.6.0_OMNIBUSF4V6.hex.zip)

Can you tell me what this version is based on? I guess you built this based on a 2.6 snapshot. 
I already have a 2.6 Build on this FC to test another feature that is already in the master branch since about 2 weeks. Is your snapshot from a later point? If not, I need to finish the other tests first before I can test the throttle optimization. If all is included I can update right away.

**Comment by avs2** [comment]
@b14ckyy I think this was branched off of master about a month ago. Here is a version (target OMNIBUSF4V6)  based on the master branch from two days ago that I flashed to my F441-WSE to test this PR. It also contains #6101 but as long as you don't touch that cli command everything will work the same as the master branch.

[inav_2.6.0_OMNIBUSF4V6.zip](https://github.com/iNavFlight/inav/files/5307431/inav_2.6.0_OMNIBUSF4V6.zip)

**Comment by Airwide** [comment]
> @b14ckyy I think this was branched off of master about a month ago. Here is a version (target OMNIBUSF4V6) based on the master branch from two days ago that I flashed to my F441-WSE to test this PR. It also contains #6101 but as long as you don't touch that cli command everything will work the same as the master branch.
> 
> [inav_2.6.0_OMNIBUSF4V6.zip](https://github.com/iNavFlight/inav/files/5307431/inav_2.6.0_OMNIBUSF4V6.zip)

@b14ckyy, @avs2 is correct. This is the last commit before my branch:

commit 19e882fea304688e271ad57c75ffd49f11012ff6
Merge: 5d221d3a8 faedca0cc
Author: Paweł Spychalski <pspychalski@gmail.com>
Date:   Fri Sep 4 12:41:12 2020 +0200

    Merge pull request #6099 from iNavFlight/dzikuvx-vscode-docs

    VSCode instructions

**Comment by b14ckyy** [comment]
Awesome. Thanks guys. I'll flash and test either tomorrow after work or on saturday.

**Comment by avs2** [comment]
I tested this yesterday with my fairly heavy dart 250 in 25km/h winds (gusting to 36km/h) and I think it works very good! The combination of a slow moving average pitch angle and a fast moving current pitch angle made for a stable throttle in level flight, but fast response when climbing or diving. When the model was bumped around by the wind it sometimes added 1 or 2 % throttle which was enough to maintain altitude.

I thought it would be good to have a better understanding of how this PR changes the throttle signal. My fc doesn't have a blackbox so I simulated a AR(1) process to model the disturbances of the actual pitch angle from the target. The situation I simulated is one where you are flying level and then pitch up 10 degrees indefinitely. This gave these results:
throttleplot

Here is the same plot without the disturbances:
throttleplot

In this plot, slowCorrection is the throttle correction for the `averagePitch` angle and `fastCorrection` is the additional correction when the difference between `pitch` and `averagePitch` becomes too large. As you can see the throttle signal is instantly increased when the plane pitches up, but there is a kink in the throttle curve when `averagePitch` catches up to `pitch` and the difference between them becomes smaller than `nav_fw_pitch2thr_threshold`. I can't say I noticed this in flight, after all this is happening within a second and the pitch variation masks it, but it is good to be aware of this. Could there bee combinations of settings where this is more noticeable?

Overall I am very happy with the throttle behavior with this PR and I think it should be considered to enable this feature by default. What you you think?

The settings I used are:
```
nav_fw_cruise_thr = 1250 # Motor overpowered for this model
nav_fw_pitch2thr = 10
nav_fw_pitch2thr_smoothing = 1024
nav_fw_pitch2thr_threshold = 30
looptime = 1000
```

**Comment by Airwide** [comment]
@avs2 Thanks for helping out, and great analysis. 

@b14ckyy How did your test go?

**Comment by Airwide** [comment]
I tested this yesterday with my Mini Talon and I believe it worked pretty good. Below is a graph showing the Throttle Output.

Screenshot 2020-10-14 at 06 10 08

My settings:
```
nav_fw_cruise_thr = 1400
nav_fw_pitch2thr = 10
nav_fw_pitch2thr_smoothing = 2048
nav_fw_pitch2thr_threshold = 50
looptime = 1000
```

**Comment by Airwide** [comment]
Now it's just a matter of finding the right settings for your use case. The good thing is that the current behaviour is kept with the default settings `nav_fw_pitch2thr_smoothing = 1`
`nav_fw_pitch2thr_threshold = 0`

From my perspective this PR is ready for merge. 

@digitalentity What do you think about the additional commits since you review?

**Comment by avs2** [comment]
> I tested this yesterday with my Mini Talon and I believe it worked pretty good. Below is a graph showing the Throttle Output.
> 
> Screenshot 2020-10-14 at 06 10 08
> 
> My settings:
> 
> ```
> nav_fw_cruise_thr = 1400
> nav_fw_pitch2thr = 10
> nav_fw_pitch2thr_smoothing = 2048
> nav_fw_pitch2thr_threshold = 50
> looptime = 1000
> ```

Do you happen to have the pitch angle and/or altitude data as well?

**Comment by Airwide** [comment]
@avs2 I've added pitch and altitude to the diagram:
Screenshot 2020-10-22 at 06 52 33
Notes:
- There are two 90 degree turns in between each altitude change, which makes the data harder to interpret.
- The difference in pitch angle for ascend and descend could be caused by wind direction and/or board alignment.
- Too make evaluation easier, ascents and descents should be made in a straight line. I'll try that next time I get up in the air.
- I'm thinking that the lack of throttle increase during the 5 degree ascend indicates that the `nav_fw_pitch2thr_smoothing = 2048` is too high. I wish we new the frequency for the throttle correction calls, that would simplify the analysis.

**Comment by kniuk** [comment]
It seems very promising. I would love to test that. @Airwide could you build it for me (Matek F765)
P.S.
I am flying BushMule.

**Comment by Airwide** [comment]
> It seems very promising. I would love to test that. @Airwide could you build it for me, please? (Matek F765 WING)
> P.S.
> I am flying BushMule.

Here you go: 

[inav_2.6.0_MATEKF765.hex.zip](https://github.com/iNavFlight/inav/files/5482620/inav_2.6.0_MATEKF765.hex.zip)

Thanks!

**Comment by kniuk** [comment]
Thank You @Airwide. I should be able to try it out before the weekend

**Comment by kniuk** [comment]
@Airwide you are refering to centidegrees (1 degree = 100 centidegrees), however I guess you meant decidegrees (1degree = 10 decidegrees)

**Comment by Airwide** [comment]
> @Airwide you are refering to centidegrees (1 degree = 100 centidegrees), however I guess you meant decidegrees (1degree = 10 decidegrees)

@kniuk You are correct! 👍 I've edited the first post. I'll post a commit eventually too.

File: src/main/navigation/navigation_fixedwing.c [hunk 5]
```diff
@@ -463,7 +463,23 @@ int16_t applyFixedWingMinSpeedController(timeUs_t currentTimeUs)
 
 int16_t fixedWingPitchToThrottleCorrection(int16_t pitch)
 {
-    return DECIDEGREES_TO_DEGREES(pitch) * navConfig()->fw.pitch_to_throttle;
+    // Calculate base throttle correction from pitch moving average
+    const int16_t movingAverageCycles = navConfig()->fw.pitch_to_throttle_smooth;
+    static int16_t averagePitch = 0;
+
+    averagePitch = (averagePitch * movingAverageCycles + pitch - averagePitch) / movingAverageCycles;
```
Comment (digitalentity) [inline src/main/navigation/navigation_fixedwing.c]: It's better to use PT1 filter here and controll the cutoff frequency (alpha). Cycles count is susceptible to the rate at which controller is updated.

File: src/main/navigation/navigation_fixedwing.c [hunk 6]
```diff
@@ -463,7 +463,23 @@ int16_t applyFixedWingMinSpeedController(timeUs_t currentTimeUs)
...
+    const int16_t baseThrottleCorrection = DECIDEGREES_TO_DEGREES(averagePitch) * navConfig()->fw.pitch_to_throttle;
+
+    // Calculate final throttle correction
+    if (pitch > (averagePitch + navConfig()->fw.pitch_to_throttle_thresh)) {
+        return baseThrottleCorrection + DECIDEGREES_TO_DEGREES(pitch - averagePitch - navConfig()->fw.pitch_to_throttle_thresh) * navConfig()->fw.pitch_to_throttle;
```
Comment (digitalentity) [inline src/main/navigation/navigation_fixedwing.c]: This essentially calculates the throttle correction based on  `pitch - fw.pitch_to_throttle_thresh`, essentially clipping the response outside of the deadband.

Wouldn't it make sense to calculate the out of deadband throttle correction based on `pitch` itself, allowing us to support rapid change in pitch with equally rapid change in throttle? Then, once the `averagePitch` catches up, the response will be smoothed again.

WDYT?

Comment (avsaase) [reply src/main/navigation/navigation_fixedwing.c]: Since `fw.pitch_to_throttle_thresh` is fixed, isn't this equivalent?

Comment (digitalentity) [reply src/main/navigation/navigation_fixedwing.c]: No, `pitch * fw.pitch_to_throttle` > `(pitch - fw.pitch_to_throttle_thresh) * fw.pitch_to_throttle`.

Your PR not only smooths the response, it also clips it considerably. Basically, the maximum possible throttle correction (`min/max_pitch * fw.pitch_to_throttle`) with this PR will never be reached initially, only after `averagePitch` catches up. 

With a large enough averaging delay and rapid enough pitch maneuver it may cause a stall in a configuration that would fly fine without this PR.

Comment (avsaase) [reply src/main/navigation/navigation_fixedwing.c]: Hmm, I see what you mean but as long as the moving average is calculated over a not too long window I highly doubt there will be problems with this PR. But I agree that a filter on the pitch angle may be a cleaner approach.

This is not my PR, btw. I just requested the feature and got involved from there. I'll look into this PT1 filter and see if I can come up with something, but perhaps @Airwide wants to do that first.

Comment (Airwide) [reply src/main/navigation/navigation_fixedwing.c]: @digitalentity Thanks for you feedback The calculation as it is right now, will make a _partial_ throttle compensation for pitch deviations outside the deadband, until the moving average catches up.

File: src/main/navigation/navigation_fixedwing.c (hunk 5 above)
Comment (Airwide) [reply src/main/navigation/navigation_fixedwing.c]: @digitalentity - Thank you for pointing me in the PT1 direction. I'll take a look at it. Do you know if it's possible to filter below 1 Hz?

File: src/main/navigation/navigation_fixedwing.c (hunk 6 above)
Comment (kniuk) [reply src/main/navigation/navigation_fixedwing.c]: Hi, I did test the behaviour of threshold averaging by @Airwide. It works quite fine. Helps a lot with constantly waving rpms on my BushMule.
I noticed that when you switch off averaging (setting smoothness to 1)  threshold has no effect.

Regards,
Jacek

Comment (digitalentity) [reply src/main/navigation/navigation_fixedwing.c]: @Airwide  I don't like the **partial** part of this PR. I like the deadband and smoothing the response when it's not necessary, but please change the logic to allow the throttle compensation to use the full extent of the throttle compensation in the extremes.

File: src/main/navigation/navigation_fixedwing.c (hunk 5 above)
Comment (digitalentity) [reply src/main/navigation/navigation_fixedwing.c]: For our PT1 filter implementation the cutoff frequency can be fractional, so no issue with having a sub-Hz filters.

File: src/main/navigation/navigation_fixedwing.c (hunk 6 above)
Comment (Airwide) [reply src/main/navigation/navigation_fixedwing.c]: @digitalentity Changed to full throttle compensation outside of deadband.

File: src/main/navigation/navigation_fixedwing.c (hunk 5 above)
Comment (Airwide) [reply src/main/navigation/navigation_fixedwing.c]: @digitalentity I have now implemented a PT1 filter similar to the one for control smoothness.

**Comment by Airwide** [comment]
This PR has been changed after feedback from @digitalentity. 

- A low pass filter has replaced the moving average. Setting `fw.pitch_to_throttle_smooth` between 0 and 9 adjust the cutoff frequency between `NAV_FW_BASE_PITCH_CUTOFF_FREQUENCY_HZ` = 2 Hz and 0.1 Hz according to the following:
  - `fw.pitch_to_throttle_smooth = 0`: 2.1 Hz
  - `fw.pitch_to_throttle_smooth = 1`: 1.6 Hz
  - `fw.pitch_to_throttle_smooth = 2`: 1.1 Hz
  - `fw.pitch_to_throttle_smooth = 3`: 0.79 Hz
  - `fw.pitch_to_throttle_smooth = 4`: 0.53 Hz
  - `fw.pitch_to_throttle_smooth = 5`: 0.35 Hz
  - `fw.pitch_to_throttle_smooth = 6`: 0.23 Hz
  - `fw.pitch_to_throttle_smooth = 7`: 0.15 Hz
  - `fw.pitch_to_throttle_smooth = 8`: 0.12 Hz
  - `fw.pitch_to_throttle_smooth = 9`: 0.10 Hz

- Changed to full throttle compensation outside of the deadband defined by `fw.pitch_to_throttle_thresh`.
- Default values of `fw.pitch_to_throttle_smooth = 0` and `fw.pitch_to_throttle_thresh = 0` keeps current behaviour.

@digitalentity , @avsaase , @kniuk , @b14ckyy WDYT?

I plan to head out and test tomorrow, if the weather is ok.

**Comment by avsaase** [comment]
Awesome! 

I think this way of filtering the pitch angle is better, and the configuration much easier. I do wonder if we still need the threshold and if we can't just apply the filter to all pitch angles. As long as filteredPitch is not too sluggish and cruise throttle is set at a safe level I don't think that will ever cause problems. The threshold also causes a weird kink in the throttle response when climbing or diving at an angling outside of the threshold, for example clinbing at 10 degrees when `nav_fw_pitch_to_throttle_thresh = 30`:
Rplot
 
This plot still uses averagePitch instead of filteredPitch, but I think the results would be similar. As filteredPitch is catching up to pitch, the difference becomes smaller than the threshold so there is an abrupt changes to using filteredPitch instead of pitch for the throttle correction calculation. Since filteredPitch has not yet fully caught there is a drop in the throttle output. I think any use of an if statement will always cause some sort of kink in the throttle response so I propose to just get rid of it. If @digitalentity doesn't want that an alternative would be to increase the max value for `pitch_to_throttle_thresh` to 900 so that informed users can disable it themselves.

Weather permitting I will test this weekend. What would be a good starting point for `nav_fw_pitch_to_throttle_smooth`?

PS.: GitHub indicates some line changes, but all that is different is the line ending?

**Comment by Airwide** [comment]
@avsaase I'm aware of the weird kink, and that's the reason why I implemented it the way I did in the first place with partial compensation outside the deadband. However, I think @digitalentity has a valid point about full compensation to avoid the risk of stalls. Let's try this version and see if the kink is a problem in reality.

Interesting that you are thinking about a not too sluggish filter. I'm thinking the other way around to get a really stable throttle. In fact, I'm not sure `nav_fw_pitch2thr_smoothing = 9` which equals 0.10 Hz is low enough. I'm thinking of altering the cutoff frequency setting to enable an even lower frequency according to the below graph:

Screenshot 2020-11-13 at 10 55 23

I could increase `nav_fw_pitch2thr_threshold` maximum value at the same time to facilitate testing of a wide range of different setup configurations. Once we find a good default setting, we could narrow down the interval if necessary.

WDYT?

Btw, I went out for testing yesterday. Both my Dart 250 and my Mini Talon sky rocketed at the start of the test mission, so I had to abort the testing. I found out later that I had a 0 too many in the altitude value for all the waypoints. 👎 I will head out again as soon as the weather permits.

**Comment by Airwide** [comment]
> PS.: GitHub indicates some line changes, but all that is different is the line ending?

Sorry about that. I have automatic removal of line end whitespace at save in one of my code editors.

**Comment by avsaase** [comment]
@Airwide I am not too familiar with low-pass filters so I can't really comment on sensible cutoff frequencies. I think the slower the `filteredPitch`, the longer the kink in the throttle curve so that is becomes more noticeable.

**Review by digitalentity — APPROVED** [review]
Looks good! Did you have a chance to test it yet?

**Comment by Airwide** [comment]
@digitalentity I did get one flight in yesterday with the Dart 250 before the rain. I used the following settings:

```
nav_fw_pitch2thr = 22
nav_fw_pitch2thr_smoothing = 6
nav_fw_pitch2thr_threshold = 50
```

Throttle correction was smoothed a bit. Immediate full correction outside of deadband worked as expected. No signs of the "kink" during transitions by looking at the dvr footage.

I'm going to push another commit with an increased span of cutoff frequency beyond 0.1 Hz and also open up `nav_fw_pitch2thr_threshold` max value to give @avsaase a chance to test his ideas.

My next opportunity for testing is Tuesday if I'm lucky. I plan to test with much more smoothing and hopefully both with the Dart250 and the Mini Talon.

@kniuk, @b14ckyy - I could build hex files for you, if you are interested in helping out with testing.

**Comment by digitalentity** [comment]
I'm going to merge this now. I think we can keep the settings as-is - setting `pitch_to_throttle_thresh` to a large value (900) will effectively always apply the LPF which should be reasonably safe in most cases. This needs to be tested, but it's not a blocker for this PR. Thanks for looking into this feature!

**Comment by Airwide** [comment]
@digitalentity I'm sorry, I just realized I did a mistake in the last commit. I accidentally changed `getSmoothnessCutoffFreq()` function instead of `getPitchToThrottleSmoothnessCutoffFreq` function. I made another commit to fix it, but the PR is already merged. Do I need to file another PR?

**Comment by avsaase** [comment]
Maybe we should have some documentation on this feature? In fact, when we find sensible values I think it should be considered to enable it by default (and/or add it to the configurator). There are already several features that make INAV fly better but are hidden in the CLI where most inexperienced users will never find them.  In my opinion this shouldn't be one of them.

**Comment by Airwide** [comment]
@avsaase Agreed! Let's try and find a good basic setup together and then change the default value in a separate PR.

Btw, I did get a test flight in this morning with my Mini Talon. I must say it is working good.
```
Settings:
nav_fw_cruise_thr = 1400
nav_fw_control_smoothness = 0
nav_fw_pitch2thr = 10
nav_fw_pitch2thr_smoothing = 9 // Note that this was with the old smoothing calculation where 9 equals 0.1 Hz
nav_fw_pitch2thr_threshold = 50
```
Here are two screenshots showing the behaviour with pitch changes outside of the deadband.

Screenshot 2020-11-17 at 10 56 32

Screenshot 2020-11-17 at 11 23 00

I want to test with an even lower frequency. It would also be interesting to combine these settings with some`nav_fw_control_smoothness`. I think that could be an alternative to increasing `nav_fw_pitch2thr_threshold`.

**Comment by avsaase** [comment]
I finally found the time and the weather to test this, and works very good! My setting were:
```
nav_fw_cruise_thr = 1250
nav_fw_pitch2thr = 10
nav_fw_pitch2thr_smoothing = 4
nav_fw_pitch2thr_threshold = 50
nav_fw_control_smoothness = 4 // I didn't know I had this setting enabled
```
This is with the implementation in RC3. Is that with the new smoothing curve? Not that it makes a big difference with these settings. 

I think these settings for `nav_fw_pitch2thr_smoothing` and `nav_fw_pitch2thr_threshold` are generally safe to use and won't cause problems, so I propose making them the new defaults.

**Comment by Airwide** [comment]
> This is with the implementation in RC3. Is that with the new smoothing curve? 

@avsaase Yes, RC3 has the new smoothing curve.

> I think these settings for `nav_fw_pitch2thr_smoothing` and `nav_fw_pitch2thr_threshold` are generally safe to use and won't cause problems, so I propose making them the new defaults.

Your suggestion for default values is quite conservative, which can be a good thing. However, I would consider `nav_fw_pitch2thr_smoothing = 6` together with `nav_fw_pitch2thr_threshold = 50` for defaults.

@digitalentity WDYT?
